### DISCUSSION TITLE
How to setup micropython in Terminal

### DISCUSSION THREAD

**Discussion by heckmon** [micropython/micropython#13722] [General]
I am new to micropython, Currently I am using it in Thonny ide.
Is there any way to install it in the terminal (or command prompt) like standard python?

**Comment by Jeff-Squared** [comment]
Take a look at mpremote https://docs.micropython.org/en/latest/reference/mpremote.html

**Reply by heckmon** [reply]
I thought mpremote was mainly used for installing MicroPython packages via mip, but this is actually cool mpremote has lots of stuff. It seems I can program any board with mpremote.

Thanks a lot!

**Comment by glenn20** [comment]
> I am new to micropython, Currently I am using it in Thonny ide.
> Is there any way to install it in the terminal (or command prompt) like standard python?

As suggested by @Jeff-Squared , mpremote is a great way to work with the micropython repl from within a terminal/cli.

To install micropython onto your device from the command line, check the install instructions for your device (you didn't mention which board type you are using) on the micropython downloads page (https://micropython.org/download/). For and esp32 device, esptool is your friend.

**Reply by heckmon** [reply]
Thanks a lot!
I am currently working with ESP32-WROOM-32. I installed esptools and it's working well

**Comment by shariltumin** [comment]
If you like mpremote you probably also like ```mpy-cross``` and ```mpy-tool```.

mpy-cross:
```sh
$ mpy-cross --help
usage: mpy-cross [<opts>] [-X <implopt>] [--] <input filename>
Options:
--version : show version information
-o : output file for compiled bytecode (defaults to input filename with .mpy extension, or stdout if input is stdin)
-s : source filename to embed in the compiled bytecode (defaults to input file)
-v : verbose (trace various operations); can be multiple
-O[N] : apply bytecode optimizations of level N

Target specific options:
-msmall-int-bits=number : set the maximum bits used to encode a small-int
-march=<arch> : set architecture for native emitter; x86, x64, armv6, armv6m, armv7m, armv7em, armv7emsp, armv7emdp, xtensa, xtensawin

Implementation specific options:
  emit={bytecode,native,viper} -- set the default code emitter
  heapsize=<n> -- set the heap size for the GC (default 2097152)
```

mpy-tool
```sh
$ mpy-tool -h
usage: mpy-tool [-h] [-x] [-d] [-f] [--merge] [-q QSTR_HEADER]
                [-mlongint-impl {none,longlong,mpz}] [-mmpz-dig-size N]
                [-o OUTPUT]
                files [files ...]

A tool to work with MicroPython .mpy files.

positional arguments:
  files                 input .mpy files

options:
  -h, --help            show this help message and exit
  -x, --hexdump         output an annotated hex dump of files
  -d, --disassemble     output disassembled contents of files
  -f, --freeze          freeze files
  --merge               merge multiple .mpy files into one
  -q QSTR_HEADER, --qstr-header QSTR_HEADER
                        qstr header file to freeze against
  -mlongint-impl {none,longlong,mpz}
                        long-int implementation used by target (default mpz)
  -mmpz-dig-size N      mpz digit size used by target (default 16)
  -o OUTPUT, --output OUTPUT
                        output file
```

**Reply by heckmon** [reply]
I guess these tools convert python code to bytecodes, like in C right? So it saves memory.

So these tools are useful in resource constrained microcontrollers right?

Please correct me if I am wrong,
I am completely new to this field.

I am a first-year engineering student with an interest in microcontrollers and electronics. While I don't know embedded C, I am familiar with basic Python. Recently, I came across Micropython, a language used for programming microcontrollers using Python. Consequently, I purchased an ESP32-WROOM to delve into Micropython. I've experimented with basic tasks like LED blinking and seven-segment display control.

Unfortunately, there is a limited amount of resources about Micropython available on YouTube.

Could you provide a link to any courses or videos that cover complete Micropython learning?

**Reply by shariltumin** [reply]
You are right, these tools help us to develop programs. I will use esp8266 as an example. Suppose we have a script "pow.py".

```py
import math

def t(n,m):
    return math.sqrt(math.pow(n,m))

print(t(2,3))
```

We can compile "pow.py" (source) to "pow.mpy" (bytecode)

```sh
$ mpy-cross -o ESP8266/pow.mpy -march=xtensa pow.py
```

We can then have a look at the compiled "pow.mpy"

```sh
$ cd ESP8266
$ mpy-tool -x pow.mpy
00000000: 4d06 001f 0900 0c70 6f77 2e70 7900 0f08  M......pow.py...  pow.py <module>
00000010: 6d61 7468 0002 7400 0873 7172 7400 8175  math..t..sqrt..u  math t sqrt pow
00000020: 8177 026e 0002 6d00 8164 1806 0146 6480  .w.n..m..d...Fd.  print n m <module>
00000030: 511b 0216 0232 0016 0311 0611 0382 8334  Q....2.........4 
00000040: 0234 0159 5163 0181 283a 0803 0708 6012  .4.YQc..(:....`.  t
00000050: 0214 0412 0214 05b0 b136 0236 0163       .........6.6.c 
```

We are also interested in understanding what these hex numbers represent

```sh
$ mpy-tool -d pow.mpy
mpy_source_file: pow.mpy
source_file: pow.py
header: 4d:06:00:1f
qstr_table[9]:
    pow.py
    <module>
    math
    t
    sqrt
    pow
    print
    n
    m
obj_table: []
simple_name: <module>
  raw bytecode: 28 18:06:01:46:64:80:51:1b:02:16:02:32:00:16:03:11:06:11:03:82:83:34:02:34:01:59:51:63
  prelude: (4, 0, 0, 0, 0, 0)
  args: []
  line info: 46:64
  80          LOAD_CONST_SMALL_INT 0 
  51          LOAD_CONST_NONE 
  1b:02       IMPORT_NAME math
  16:02       STORE_NAME math
  32:00       MAKE_FUNCTION 0
  16:03       STORE_NAME t
  11:06       LOAD_NAME print
  11:03       LOAD_NAME t
  82          LOAD_CONST_SMALL_INT 2 
  83          LOAD_CONST_SMALL_INT 3 
  34:02       CALL_FUNCTION 2
  34:01       CALL_FUNCTION 1
  59          POP_TOP 
  51          LOAD_CONST_NONE 
  63          RETURN_VALUE 
  children: ['t']
simple_name: t
  raw bytecode: 21 3a:08:03:07:08:60:12:02:14:04:12:02:14:05:b0:b1:36:02:36:01:63
  prelude: (8, 0, 0, 2, 0, 0)
  args: ['n', 'm']
  line info: 60
  12:02       LOAD_GLOBAL math
  14:04       LOAD_METHOD sqrt
  12:02       LOAD_GLOBAL math
  14:05       LOAD_METHOD pow
  b0          LOAD_FAST 0 
  b1          LOAD_FAST 1 
  36:02       CALL_METHOD 2
  36:01       CALL_METHOD 1
  63          RETURN_VALUE 
  children: []
```

As you can see, the decompiler shows quite a lot of information. And finally, we want to run our program

```sh
$ mpremote u0 mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyUSB0
Use Ctrl-] or Ctrl-x to exit this shell
>
MicroPython v1.22.0-preview.246.g4365edb81.dirty on 2024-01-15; WEMOS D1 MINI 4MB with ESP8266
>>> import pow
2.82843
>>> 
```

The "pow.mpy" was not copied to flash, mpremote mounts our "ESP8266" to "/remote" on the development board.  This prolongs the life of the flash memory. You can copy the "pow.mpy" file to the flash later after development and testing is complete.

I find these URLs helpful:
1. https://docs.micropython.org/en/latest/
2. https://github.com/micropython/micropython/tree/master/tests
3. https://github.com/micropython/micropython/tree/master/examples

You have in your hands inexpensive yet powerful tools for learning microcontroller programming. MicroPython opens up a world of possibilities. Good luck, my best wishes for your success.

**Reply by heckmon** [reply]
Very useful information.
Thanks a lot Sir.☺️

Executing a program on a microcontroller without the need to upload it to the flash memory is quite impressive.

**Comment by DavesCodeMusings** [comment]
If you find yourself using VS Code, I have [a couple extensions](https://marketplace.visualstudio.com/publishers/DavesCodeMusings) that might interest you. They are wrappers around the esptool and mpremote Python modules to give you point and click access to the tools from within VS Code.

**Reply by heckmon** [reply]
VScode is my favourite editor. I used to program my ESP32 with VScode and the Platform io extension.

But your extensions are really nice sir, 
The ESPtool extension is very much useful,
It really saves time.
